---
title: "markdowntemplate"
author: "Dan"
date: "2/28/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#list of packages to install
cran_pkg <- c("BiocManager", "bookdown", "dplyr", "ecodist", "ggplot2", 
              "gridExtra", "kableExtra", "knitr", "scales", "vegan")
bioc_pkg <- c("ANCOMBC", "ape", "DESeq2",  "DirichletMultinomial", "mia", "miaViz")

#Lists of packages alrdy installed
cran_pkg_already_installed <- cran_pkg[ cran_pkg %in% installed.packages() ]
bioc_pkg_already_installed <- bioc_pkg[ bioc_pkg %in% installed.packages() ]

#Lists of packages to install
cran_pkg_to_be_installed <- setdiff(cran_pkg, cran_pkg_already_installed)
bioc_pkg_to_be_installed <- setdiff(bioc_pkg, bioc_pkg_already_installed)

#command to get packages to install from CRAN
if( length(cran_pkg_to_be_installed) ) {
  install.packages(cran_pkg_to_be_installed)
}

#command to get packages to install from bioc
if( length(bioc_pkg_to_be_installed) ) {
  BiocManager::install(bioc_pkg_to_be_installed, ask = F)
}

#reorder packages to prioritize mia and miaViz (since some functions overlap)
bioc_pkg <- c(bioc_pkg[ bioc_pkg %in% c("mia", "miaViz") ], 
              bioc_pkg[ !bioc_pkg %in% c("mia", "miaViz") ] ) 

# Loading all packages into session. Returns true if package was successfully loaded.
loaded <- sapply(c(bioc_pkg, cran_pkg), require, character.only = TRUE)
as.data.frame(loaded)
```

## [importing data into R](https://microbiome.github.io/course_2021_radboud/importing-microbiome-data.html)

## Lesson 1: objectives

ADHD Data was imported, data used [textbook](https://microbiome.github.io/OMA/data-introduction.html#loading-experimental-microbiome-data) as guidelines. 

Refer to reports/example_import for exemplary solution

## Background on datafile types
Data imported and presented as summarizedexperiment object.
Summarizedexperiment contains the following:
- assays : similar to otu_table in phyloseq. In SummarizedExperiment object, multiple assays, raw counts, transformed counts can be stored. See also MultiAssayExperiment for storing data from multiple experiments such as RNASeq, Proteomics, etc. 

- rowData : similar to tax_table in phyloseq, stores taxonomic information. 

- colData : similar to sample_data in phyloseq, stores information related to samples. 

- rowTree : similar to phy_tree in phyloseq, stores phylogenetic tree.

- FeatureIDs : OTU/ASV ids, rownames in assays and rowData

- SampleIDs: sample IDs, column names in assays, row names in colData



```{r importing data}
#data is being imported in Biom, csv and tre format; biom contains abundance table and taxonomy information; csv contains sample metadata, tree file contains phylogenetic tree

#first create objects for their file paths
biom_path <- 'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Aggregated_humanization2.biom'
csv_path <-  'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Mapping_file_ADHD_aggregated.csv'
tre_path <- 'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Data_humanization_phylo_aggregation.tre'

#load biom file 
se <- loadFromBiom(biom_path)

#asssays in se gives a list of abundance tables, which are named as 'counts'
assays(se)$counts[1:3, 1:3]

#the rowdata gives taxonomic information
head(rowData(se))

#rowData is messy because the column names (taxonomy1-6) are not real names, they should be different taxonomic hierarchies; also all taxa include 'k___', 'p____', so remove for clarity
names(rowData(se)) <- c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus')

#use bplapply to go through whole dataframe and remove '.*[kpocfg]__' from strings, where [kpocfg] is any character listed inside the [] and .* refers to any character
rowdata_modified <- BiocParallel::bplapply(rowData(se), FUN=stringr::str_remove,
                                           pattern='.*[kpcofg]__')

#Genus level has a '\"' (i.e. single quotation marks) at the end, so we'll delete that too, note the forward slash needed to mark the quotation marks
rowdata_modified <- BiocParallel::bplapply(rowdata_modified, 
                                           FUN=stringr::str_remove,
                                           pattern='\"')

#convert into dataset
rowdata_modified <- DataFrame(rowdata_modified)

#Assign rowdata_modified back to se object
rowData(se) <- rowdata_modified

#now the table is cleaned up
head(rowData(se))

#the Biom file does not contain sample metadata, so it has empty dataframe
head(colData(se))

#So, have to import sample metadata from the csv file earlier 
read.table(csv_path)

#note the csv has no headers, also need to convert from data.frame to DataFrame
sample_meta <- DataFrame(read.table(csv_path, sep=',', header=FALSE))

#use the first column of the csv file as the rownames
rownames(sample_meta) <- sample_meta[,1]

#delete sample names itself, so the 1st column is now official rownames and not part of the csv file itself
sample_meta[,1] <- NULL

#assign headers
colnames(sample_meta) <- c('patient_status', 'cohort', 'patient_status_vs_cohort','sample_name')

#add sample_meta to colData in se
colData(se) <- sample_meta
head(colData(se))

#Next, add phylogenetic tree - current SummarizedExperiment object ('se') does not have a slot for adding phylogenetic tree, so we need to add a rowTree slot

tse <- as(se, 'TreeSummarizedExperiment')

#tse now has the same data as se
tse

#read tree file and assign it to rowTree in tse
tree <- ape::read.tree(tre_path)
rowTree(tse) <- tree
tse

head(rowTree(tse))

```

