---
title: "markdowntemplate"
author: "Dan"
date: "2/28/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#list of packages to install
cran_pkg <- c("BiocManager", "bookdown", "dplyr", "ecodist", "ggplot2", 
              "gridExtra", "kableExtra", "knitr", "scales", "vegan")
bioc_pkg <- c("ANCOMBC", "ape", "DESeq2",  "DirichletMultinomial", "mia", "miaViz")

#Lists of packages alrdy installed
cran_pkg_already_installed <- cran_pkg[ cran_pkg %in% installed.packages() ]
bioc_pkg_already_installed <- bioc_pkg[ bioc_pkg %in% installed.packages() ]

#Lists of packages to install
cran_pkg_to_be_installed <- setdiff(cran_pkg, cran_pkg_already_installed)
bioc_pkg_to_be_installed <- setdiff(bioc_pkg, bioc_pkg_already_installed)

#command to get packages to install from CRAN
if( length(cran_pkg_to_be_installed) ) {
  install.packages(cran_pkg_to_be_installed)
}

#command to get packages to install from bioc
if( length(bioc_pkg_to_be_installed) ) {
  BiocManager::install(bioc_pkg_to_be_installed, ask = F)
}

#reorder packages to prioritize mia and miaViz (since some functions overlap)
bioc_pkg <- c(bioc_pkg[ bioc_pkg %in% c("mia", "miaViz") ], 
              bioc_pkg[ !bioc_pkg %in% c("mia", "miaViz") ] ) 

# Loading all packages into session. Returns true if package was successfully loaded.
loaded <- sapply(c(bioc_pkg, cran_pkg), require, character.only = TRUE)
as.data.frame(loaded)
```

## [Lesson 1: importing data into R](https://microbiome.github.io/course_2021_radboud/importing-microbiome-data.html)

## Lesson 1: objectives

ADHD Data was imported, data used [textbook](https://microbiome.github.io/OMA/data-introduction.html#loading-experimental-microbiome-data) as guidelines. 

Refer to reports/example_import for exemplary solution

## Background on datafile types
Data imported and presented as summarizedexperiment object.
Summarizedexperiment contains the following:
- assays : similar to otu_table in phyloseq. In SummarizedExperiment object, multiple assays, raw counts, transformed counts can be stored. See also MultiAssayExperiment for storing data from multiple experiments such as RNASeq, Proteomics, etc. 

- rowData : similar to tax_table in phyloseq, stores taxonomic information. 

- colData : similar to sample_data in phyloseq, stores information related to samples. 

- rowTree : similar to phy_tree in phyloseq, stores phylogenetic tree.

- FeatureIDs : OTU/ASV ids, rownames in assays and rowData

- SampleIDs: sample IDs, column names in assays, row names in colData

### *Note that all embedded chunks will be hidden, unhide them (include=TRUE) if you want to see them for info!*

```{r importing data, include=FALSE}
#data is being imported in Biom, csv and tre format; biom contains abundance table and taxonomy information; csv contains sample metadata, tree file contains phylogenetic tree

#first create objects for their file paths
biom_path <- 'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Aggregated_humanization2.biom'
csv_path <-  'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Mapping_file_ADHD_aggregated.csv'
tre_path <- 'C:/Users/Dan/OneDrive - The University Of Hong Kong/Desktop/intro_microbiome_data/data/Data_humanization_phylo_aggregation.tre'

#load biom file 
se <- loadFromBiom(biom_path)

#asssays in se gives a list of abundance tables, which are named as 'counts'
assays(se)$counts[1:3, 1:3]

#the rowdata gives taxonomic information
head(rowData(se))

#rowData is messy because the column names (taxonomy1-6) are not real names, they should be different taxonomic hierarchies; also all taxa include 'k___', 'p____', so remove for clarity
names(rowData(se)) <- c('Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus')

#use bplapply to go through whole dataframe and remove '.*[kpocfg]__' from strings, where [kpocfg] is any character listed inside the [] and .* refers to any character
rowdata_modified <- BiocParallel::bplapply(rowData(se), FUN=stringr::str_remove,
                                           pattern='.*[kpcofg]__')

#Genus level has a '\"' (i.e. single quotation marks) at the end, so we'll delete that too, note the forward slash needed to mark the quotation marks
rowdata_modified <- BiocParallel::bplapply(rowdata_modified, 
                                           FUN=stringr::str_remove,
                                           pattern='\"')

#convert into dataset
rowdata_modified <- DataFrame(rowdata_modified)

#Assign rowdata_modified back to se object
rowData(se) <- rowdata_modified

#now the table is cleaned up
head(rowData(se))

#the Biom file does not contain sample metadata, so it has empty dataframe
head(colData(se))

#So, have to import sample metadata from the csv file earlier 
read.table(csv_path)

#note the csv has no headers, also need to convert from data.frame to DataFrame
sample_meta <- DataFrame(read.table(csv_path, sep=',', header=FALSE))

#use the first column of the csv file as the rownames
rownames(sample_meta) <- sample_meta[,1]

#delete sample names itself, so the 1st column is now official rownames and not part of the csv file itself
sample_meta[,1] <- NULL

#assign headers
colnames(sample_meta) <- c('patient_status', 'cohort', 'patient_status_vs_cohort','sample_name')

#add sample_meta to colData in se
colData(se) <- sample_meta
head(colData(se))

#Next, add phylogenetic tree - current SummarizedExperiment object ('se') does not have a slot for adding phylogenetic tree, so we need to add a rowTree slot

tse <- as(se, 'TreeSummarizedExperiment')

#tse now has the same data as se
tse

#read tree file and assign it to rowTree in tse
tree <- ape::read.tree(tre_path)
rowTree(tse) <- tree
tse

head(rowTree(tse))

```

## [Lesson 2: exploring data](https://microbiome.github.io/course_2021_radboud/microbiome-data-exploration.html)

## Lesson 2: objectives
- Microbial abundances are relative, so we need to calculate them; they are easier to interpret than plain counts; refer to [this](https://microbiome.github.io/mia/reference/transformCounts.html) to find out the ways to transform counts

- You can also agglomerate (sum up all data) by taxonomic rank

### Part 1: Exploring data structure
```{r exploring data structures and manipulation, include=FALSE}

#Calculating relative abundances and putting it in assays; tse now has count and relative abundance information stored in assays
tse <- transformCounts(tse, method='relabundance')

#summarizing data by taxonomic rank
tse_phylum <- agglomerateByRank(tse, rank='Phylum')
dim(tse_phylum)

#tse_phylum shows there are 5 different phyla groups, e.g. from RowData we see all Firmicutes are combined together

knitr::kable(head(rowData(tse_phylum))) %>%
  kableExtra::kable_styling('striped', 
                           latex_options='scale_down') %>%
  kableExtra::scroll_box(width='100%')

#agglomeration is missing abundances for certain taxa, but can use 'na.rm = TRUE' to remove taxa that do not have information in specified levels

#alternatively can aggregate by a rank, then print out/select taxa that do not have info at the genus level, and note that if they don't have information at a specified level, they are aggregated at the lowest possible level left after agglomeration
temp <- rowData(agglomerateByRank(tse,rank='Genus'))

knitr::kable(head(temp[temp$Genus=='',])) %>%
  kableExtra::kable_styling('striped',
                            latex_options='scale_down') %>%
  kableExtra::scroll_box(width='100%')

temp2 <- rowData(agglomerateByRank(tse, rank='Genus', na.rm=TRUE))
print(paste0('Agglomeration with na.rm = FALSE: ', dim(temp)[1], 'taxa.'))
print(paste0('Agglomeration with na.rm = TRUE: ', dim(temp2)[1], 'taxa.'))
temp

#assays are a list of matrices, and you can access individual assays
assay(tse, 'counts')[1:5, 1:7]

#access phylogenetic trees with rowTree()

#the links to individual features made available through rowLinks
rowLinks(se)

#can convert SE data into long data frame via meltAssay, useful for data manipulation and visualization via tidyverse
molten_tse <- meltAssay(tse,
                        add_row_data=TRUE,
                        add_col_data=TRUE,
                        abund_values='relabundance')
molten_tse

#Can also subset data, will show example with data GlobalPatterns

#store data into se and check dimensions
data("GlobalPatterns", package='mia')
se2 <- GlobalPatterns
dim(se2)

#Subset and extract data only from HUMAN SAMPLES, this informaiton is stored as SampleType in colData(se), and se

#firstidentify the unique possible sampleType
unique(se2$SampleType)

#show how many of each sample type
se2$SampleType %>% table()

#subset by sample, %in% makes sure to include everything from that list in that category
se2_subset_by_sample <- se2[ ,se2$SampleType %in% c('Feces', 'Skin', 'Tongue')]
dim(se2_subset_by_sample)

#note you can subset by origin, sampling time, sequencing method, DNA/RNA barcode and cohort

#Next task is to subset by Phyla containing 'Actinobacteria' and 'Chlamydiae'
#first inspect unique values
unique(rowData(se2)$Phylum)
rowData(se2)$Phylum %>% table()

#after subsetting, if you want to find abundancies related to Actinobnact and Chlamydiae separately, then you don't need agglomeration; if you want to find abundances to both, then you need agglomeration.

#note the following object selects features with the 2 aforementioned phyla and filters out features that have NA in place of Phylum

se2_subset_by_feature <- se2[rowData(se2)$Phylum %in% c('Actinobacteria', 'Chlamydiae') & !is.na(rowData(se2)$Phylum),]

dim(se2_subset_by_feature)

se2_subset_by_feature

#Can also agglomerate by Phyla to find out total abundances of certain Phyla
se2_phylum <- se2 %>% agglomerateByRank(rank='Phylum')

se2_phylum_subset_by_feature <- se2_phylum[rowData(se2_phylum)$Phylum %in% c('Actinobacteria', 'Chlamydiae') & !is.na(rowData(se2_phylum)$Phylum),]

dim(se2_phylum_subset_by_feature)

#you can subset by feature for taxonomic rank and meta-taxonomic group
#agglomeration merged all the features with the same phlum into one big category to find the total abundance 

#You can also subset sample and feature at the same time, such as here, where we subset human samples, with phyla actinobact and chlamydiae

se2_subset_by_sample_feature <- se2[rowData(se2)$Phylum %in% c('Actinobacteria', 'Chlamydiae') & !is.na(rowData(se2)$Phylum), se2$SampleType %in% c('Feces', 'Skin', 'Tongue')]

dim(se2_subset_by_sample_feature)

```

### Part 2: Exploring data visualization

Explore [miaViz](https://microbiome.github.io/miaViz/articles/miaViz.html) to explore the different options for data visualization. 

```{r exploring data visualization, include=FALSE}
#specify 'relabundance' to be the abundance table used for plotting
#agglomeration is a built-in option, so can use agglomerated or non agglomerated tse as input

plot_abundance <- plotAbundance(tse, abund_values='relabundance', rank='Phylum')+
  theme(legend.key.height=unit(0.5, 'cm')) + 
  scale_y_continuous(label=scales::percent)

plot_abundance

#can also do density plots to show ovferall abundance distribution for a taxonomic group, like relative abundance of Firmicutes

#first subset data w/ only Firmicutes
tse_firmicutes <- tse_phylum['Firmicutes']

#retrieve abundance table
abundance_firmicutes <- assay(tse_firmicutes, 'relabundance')

#create dataframe object, with first column being abundances
firmicutes_abund_df <- as.data.frame(t(abundance_firmicutes))
firmicutes_abund_df

#rename column
colnames(firmicutes_abund_df) <- 'abund'

#create plot
firmicutes_abund_plot <- firmicutes_abund_df %>%
  ggplot(aes(x=abund))+
  geom_density(color='darkred', fill='lightblue')+
  labs(x='relative abundance', title='Firmicutes')+
  theme_classic()+
  scale_x_continuous(label=scales::percent)


firmicutes_abund_plot
```

### Part 3: Exercises 
